### PR TITLE
Codes for ODLA_POPART pipeline supports

### PR DESCRIPTION
**Pull request by jackzipu**
Very beginning of codes with some hard coded codes to run mlperf model on 1 IPU in BS6 and got the expected result.

### CODE REVIEW

File: ODLA/platforms/odla_popart/odla_compute.cc [hunk 1]
```diff
@@ -94,210 +58,236 @@ odla_status odla_SetComputationItem(odla_computation comp, odla_item_type type,
...
-
-  popart::StepIO stepio(inputs, outputs);
-  // Run on ipu
-  context->session->run(stepio);
+  //std::cout << "---> odla_ExecuteComputation()" << std::endl;
+  if(!context->hold("odla_ExecuteComputation"))
+    return ODLA_FAILURE;
+  comp->executor()->compute(comp, context, mode, device);
+  //std::cout << "<--- odla_ExecuteComputation()" << std::endl;
   return ODLA_SUCCESS;
```
Comment (youbeny) [inline ODLA/platforms/odla_popart/odla_compute.cc:138]: pipeline mode needs to be compatible with non-pipeline mode，consider add a flag， such as  
```cpp
if (comp->opts.enable_pipeline) {
        // do pipeline;
} else {
        // normal;
}
```

File: ODLA/platforms/odla_popart/odla_popart.h [hunk 2]
```diff
@@ -58,31 +71,77 @@ struct _odla_value {
...
+      if(hold_it)
+          m_instance->hold(); 
+      return m_instance;
+  }
+  bool m_done;
+  bool thread_complete_;
+  std::mutex m_init_mutex;
+  Execution* m_executor;
+  std::thread::id thread_id_of_holder;
+
```
Comment (youbeny) [inline ODLA/platforms/odla_popart/odla_popart.h:102]: mixed code style  m_xx / xx_ / _xx

**Comment by weimingzha0** [comment]
where is ODLA/platforms/odla_popart/json.hpp from? Is it some third-party code?

File: ODLA/platforms/odla_popart/odla_popart.h [hunk 3]
```diff
@@ -20,29 +20,42 @@
...
 #include <popart/sessionoptions.hpp>
-#include <popart/stepio.hpp>
 #include <popart/tensorinfo.hpp>
-#include <popart/voiddata.hpp>
-#include <random>
-#include <stdexcept>
-#include <string>
-#include <vector>
+
+#define g_comp _odla_computation::instance()
```
Comment (weimingzha0) [inline ODLA/platforms/odla_popart/odla_popart.h:33]: can we use "static _odla_computation* &g_comp =  _odla_computation::instance_" ?

**Comment by weimingzha0** [comment]
can we support multiple computation on different IPU?

File: ODLA/platforms/odla_popart/CMakeLists.txt [hunk 4]
```diff
@@ -51,3 +49,8 @@ if (NOT ODLA_BUILD_POPART_USE_CXX11ABI)
 endif()
 
 target_link_libraries(odla_popart PUBLIC ODLA custom_ops popart-only)
+
+target_include_directories(odla_popart PRIVATE
```
[the remark this answers is not in the document]
Comment (jackzipu) [reply ODLA/platforms/odla_popart/CMakeLists.txt:53]: We need this as the functions of pipeline stage setting need to operate the onnx model, this function is in odla_popart.cc, which compiled by this CMakeList.

File: ODLA/platforms/odla_popart/CMakeLists.txt [hunk 5]
```diff
@@ -20,17 +20,15 @@ option(ODLA_BUILD_POPART_CUSTOM_OPS "Link with Popart custom ops" ON)
...
                                odla_ops_process.cc odla_ops.cc
-                               odla_ops_custom.cc
-                )
-
+                               odla_ops_custom.cc odla_pipeline.cc
+                               odla_popart.cc popart_config.cc
+                               )
 if (NOT POPLAR_ROOT)
-  set(POPLAR_ROOT "/opt/poplar_sdk/poplar" CACHE PATH "Path of poplar root")
+  set(POPLAR_ROOT "/opt/poplar_sdk/poplar" CACHE STRING "Path of poplar root")
```
[the remark this answers is not in the document]
Comment (jackzipu) [reply ODLA/platforms/odla_popart/CMakeLists.txt:27]: Changed to PATH for this.

File: ODLA/platforms/odla_popart/odla_popart.h (hunk 3 above)
Comment (jackzipu) [reply ODLA/platforms/odla_popart/odla_popart.h:33]: We need the instance() function to do extra check of the thread id which is holding the computation object, like follows
`static _odla_computation* instance(bool hold_it = true){
      if(hold_it)
          instance_->hold(); 
      return instance_;
  }`

**Review by jackzipu — COMMENTED** [review]
> where is ODLA/platforms/odla_popart/json.hpp from? Is it some third-party code?

Yes, it's a json parser of third party to parse the pipeline configuration.  Only this one hpp codes needed to be included.

**Comment by jackzipu** [comment]
> can we support multiple computation on different IPU?
In current design & architecture, I'm afraid we can't.  You can see that all the OPs called without passed in odla_computation object or pointer.  We used a g_comp, which meant only 1 graph can be used at same time in one process, otherwise we could not tell which computation the Ops belonged.

For the SDK itself whether support multiple graphs with multi-threads, we can have a fast check if needed.  Thanks.

**Comment by weimingzha0** [comment]
> > where is ODLA/platforms/odla_popart/json.hpp from? Is it some third-party code?
> 
> Yes, it's a json parser of third party to parse the pipeline configuration. Only this one hpp codes needed to be included.

can we put it in a "third_party" directory? (e.g.  ODLA/platforms/odla_popart/custom_ops/third_party )

**Comment by jackzipu** [comment]
The code has been merged by pull request 522, close this PR.

**Pull request closed without merging**